### PR TITLE
New method starknet_simulateTransactions

### PR DESCRIPTION
**Pull request by kirugan**

### CODE REVIEW

File: rpc/simulation.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package rpc
+
+import "fmt"
+
+type SimulationFlag int
+
+const (
+	SkipValidateFlag SimulationFlag = iota + 1
+	SkipFeeChargeFlag
```
[the remark this answers is not in the document]
Comment (kirugan) [reply rpc/simulation.go:12]: Return an error when simulation flags passed

File: rpc/handlers.go [hunk 2]
```diff
@@ -1242,6 +1242,64 @@ func (h *Handler) TraceTransaction(hash felt.Felt) (json.RawMessage, *jsonrpc.Er
...
+		if err != nil {
+			// todo or InvalidParams ?
+			return nil, jsonrpc.Err(jsonrpc.InternalError, err.Error())
+		}
+
+		txns = append(txns, tx)
+		classes = append(classes, class)
+
+		paidFeesOnL1 := []*felt.Felt{paidFeeOnL1}
+		gasConsumed, traces, err := h.vm.Execute(txns, classes, header.Number, header.Timestamp,
```
Comment (omerfirmak) [inline rpc/handlers.go:1281]: Why are calling `vm.Execute()` for each txn?
Shouldnt we just call it once for the entire txn set?

**Comment by omerfirmak** [comment]
If I am not mistaken, EstimateFee is just SimulateTransactions but it doesnt return the traces.
So what I think we should do is; move the entire function body of EstimateFee to SimulateTransactions and just add traces to it.

Once that is done, we can make EstimateFee call SimulateTransactions.

File: rpc/handlers.go (hunk 2 above)
Comment (kirugan) [reply rpc/handlers.go]: My bad, fixed

**Comment by kirugan** [comment]
> If I am not mistaken, EstimateFee is just SimulateTransactions but it doesnt return the traces. So what I think we should do is; move the entire function body of EstimateFee to SimulateTransactions and just add traces to it.
> 
> Once that is done, we can make EstimateFee just a wrapper around SimulateTransactions.

Done, the minor issue with logging:
```
defer h.callAndLogErr(closer, "Failed to close state in starknet_estimateFee")
```
We need somehow pass information about which method is called either through `context.Context` or just plain string. WDYT?

**Comment by omerfirmak** [comment]
> We need somehow pass information about which method is called either through `context.Context` or just plain string. WDYT?

I dont think it is worth it. It can just print `Failed to close state in starknet_simulateTransactions`

**Review by omerfirmak — APPROVED** [review]
LGTM

**Pull request merged**